### ISSUE TITLE
Postgres: A record with that `id` already exists

### ISSUE THREAD

**Issue by nozer** [balderdashy/waterline#830]
Hi

This is the scenario. 
- I start with a clean database. I lift sails in dev mode and it creates the tables. 
- Then, I populate the tables with some data from a file containing many sql insert statements. (I run it via 3rd party postgres client  program)
- When I try to insert new data into postgres via waterline methods, I get the following error: 
` "A record with that `id` already exists"` 

 When I look at table definitions, I see this for id attribute: 
`"id integer NOT NULL DEFAULT nextval('users_id_seq'::regclass)," ` 

I don't know much about postgres (working on a new project that uses it) so I am not sure if this is waterline problem or my bad. But, why doesn't waterline/postgres automatically increase that id (like it would for auto increment column in mysql).  

I have to do this after each insert `SELECT pg_catalog.setval('users_id_seq', <max id>, true);` to fix the problem but it is so annoying to do that after each insert. 

Any solution would be appreciated.

**Comment by tjwebb** [comment]
Does this happen in other adapters? If not, please file in the sails-postgresql repository.

**Comment by nozer** [comment]
No, not in other adapters. this nextval() sequence thing is particular to postgres (I know it is not in MySQL and MSSQL). I don't know about others.

**Comment by tjwebb** [comment]
> this nextval() sequence thing is particular to postgres

Both mysql and mssql have sequences. So I'm not sure what you mean by that.

**Comment by nozer** [comment]
Yes, in mysql it is called auto_increment primary key and in postgres it is called serial. 
I am not talking about those. They work fine. 
When sails is creating my postgres models, it is using this definition for id field: 
`id integer NOT NULL DEFAULT nextval('users_id_seq'::regclass)` 
You see the nextval(...) thing over there? That `users_id_seq` must be increased each time a new record is added. Otherwise, it gets out of sync and one gets the error I specified above.  
So, after I restart sails in --dev mode, those sequence numbers are getting out of sync. So, I have to issue the  `SELECT pg_catalog.setval('users_id_seq', max(id), true) from users ` to increase sequence. 

This only happens after I start sails. If I, for example, insert repeatedly via sails waterline methods after sails is lifted, it does not cause this problem.  When sails is starting, it should update those sequence numbers. 

So, to get around this, I am issuing these commands after each lift of sails (dev mode and migrate option is set to alter)

```
SELECT pg_catalog.setval('users_id_seq', max(id), true) from users;
SELECT pg_catalog.setval('userroles_id_seq', max(id), true) from userroles;
SELECT pg_catalog.setval('roles_id_seq', max(id), true) from roles;
SELECT pg_catalog.setval('rolepermissions_id_seq', max(id), true) from rolepermissions;
SELECT pg_catalog.setval('groups_id_seq', max(id), true) from groups;
SELECT pg_catalog.setval('groupusers_id_seq', max(id), true) from groupusers;
```

Also, sorry but how do I move this issue to sails-postgres?

**Comment by particlebanana** [comment]
What it sounds like is the data you have in the import script has primary keys and so when you insert record `10` it doesn't increase the sequence for the id to reflect it's new value of `10`. A quick fix would be to remove the primary keys from the import script and then a new value will be grabbed from the sequence like normal. The waterline migrations handle this for you so that's why you don't see the error using WL.

**Comment by nozer** [comment]
Hi,

Thanks for the reply. I indeed do have primary keys in the import script. But I need them to be there. 
Because, when I am inserting the initial data into a table like `rolepermissions`, I need to know role and permission `id`s. 

However, below each insert, I also have command to insert the sequence. For example: 
```
INSERT INTO devices (id, "serialNumber", "deviceType") VALUES
    (1, 'TC3-131622471009', 'TC3'),
    (2, 'TC3-2', 'TC3'),
    (3, 'TC3-3', 'TC3'),
    (4, 'TC3G-1', 'TC3G'),
    (5, 'TC3G-2', 'TC3G');

   SELECT pg_catalog.setval('devices_id_seq', 5, true);
```
After running the script (and updating the sequences in the script), I can add data via waterline methods without problem. 

***Then***, when I restart sails (after inserting some data via sails methods) (I don't run my import script each time I start sails, I only run it once in the beginning of my development to put some data into the db; I just mentioned it here to give full picture) it gives me that `id` error I reported. 

So, waterline is doing something when it is starting and not updating the current sequences. 
I think it should run these commands when it is starting: 
```
SELECT pg_catalog.setval('<sequence name>', max(id), true) from <related table>;
```

So, in short, imagine sails started, db is populated with data, sequences are updated and even some data via sails waterline is added without problem. Then, I restart sails and I get sequence issue. I think it should update all those sequences after each load instead of me having to do it manually.

**Comment by dmarcelino** [comment]
Hi @nozer, have you found a solution for this? Should we move this to [sails-postgresql](https://github.com/balderdashy/sails-postgresql)

**Comment by nozer** [comment]
Hi

I have wrote some scripts in config/bootstrap.js file to get around the problem. Please move.

Here is the workaround

```
sails.on('lifted', function(){

     // update sequences
     //console.log("Updating db sequences (config/bootstrap.php) ");
     Db.run("SELECT c.relname FROM pg_class c WHERE c.relkind = 'S'")
     .then(function(result){
        var queries = [];
        _.each(result.rows, function(row){
           queries.push("SELECT pg_catalog.setval('"+ row.relname +
            "', max(id), true) from " + row.relname.replace('_id_seq', ''));
        });
        //console.log(queries);
        return Db.run(queries.join('; '));

     }).then(function(){
        //console.log('Done updating sequences');
     }).catch(function(err){
        console.log('config.bootstrap function err', err);
     });
    //...
});
```

**Comment by dmarcelino** [comment]
Thanks

**Comment by rgksugan** [comment]
Facing this issue in Mysql as well

**Comment by davemackintosh** [comment]
Is there any news on this? I'm getting this issue in Multicolour core with the latest versions of Waterline and the Postgres adapters.

Do I have to use @nozer's hack to fix the sequence manually post database start.
